File: nonlineal-structures/python/bstree_of_lists.py
# Luis Espino 2021
# Examples of Binary Search Tree of Lists

class NodeList:
    def __init__(self, value):
        self.value = value
        self.next = None

class List:
    def __init__(self):
        self.head = None

    def add(self, value):
        tmp = NodeList(value)
        tmp.next = self.head
        self.head = tmp

    def show(self):
        print("[", end = " ")
        tmp = self.head
        while tmp:
            print(tmp.value, end=" ")
            tmp = tmp.next
        print("]")

class Node:
    def __init__(self, value, list):
        self.value = value
        self.list  = list
        self.left  = None
        self.right = None

class BSTree:
    def __init__(self):
        self.root = None

    #add
        
    def add(self, value, list):
        self.root = self._add(value, list, self.root)
    
    def _add(self, value, list, tmp):
        if tmp is None:
            return Node(value, list)        
        elif (value>tmp.value):
            tmp.right=self._add(value, list, tmp.right)
        else:
            tmp.left=self._add(value, list, tmp.left)
        return tmp

    #traversal

    def show(self):
        self._show(self.root)

    def _show(self, tmp):
        if tmp:
            self._show(tmp.left)
            print(str(tmp.value), end = " ")
            tmp.list.show()
            self._show(tmp.right)

#init
t = BSTree()

#add
tmp = List()
tmp.add("Fernandez")
tmp.add("Santiago")
bst = BSTree()
bst.add(25, tmp)

tmp = List()
tmp.add("Lemus")
tmp.add("Karla")
bst.add(5, tmp)

tmp = List()
tmp.add("Cabrera")
tmp.add("Angel")
bst.add(10, tmp)

tmp = List()
tmp.add("Gonzalez")
tmp.add("Carlos")
bst.add(41, tmp)

#print tree
bst.show()

File: 12-other-trees/list-bstree/go/bstree-of-lists.go
// Luis Espino 2021
// Example of Binary Search Tree of Lists

package main

import "fmt"

// NodeList is a integer with next pointer
type NodeList struct {
	data string
	next *NodeList
}

// List with stack insert
type List struct {
	head *NodeList
}

func (l *List) add(data string) {
	l.head = &NodeList{data: data, next: l.head}
}

func (l *List) show() {
	tmp := l.head
	fmt.Print("[ ")
	for tmp != nil {
		fmt.Print(tmp.data, " ")
		tmp = tmp.next
	}
	fmt.Println("]")
}

// Node is a key with two pointers
type Node struct {
	key   int
	value *List
	left  *Node
	right *Node
}

// BST is a set of sorted Nodes
type BST struct {
	root *Node
}

func (bst *BST) add(key int, value *List) {
	bst.root = bst._add(key, value, bst.root)
}

func (bst BST) _add(key int, value *List, tmp *Node) *Node {
	if tmp == nil {
		return &Node{key: key, value: value}
	} else if key > tmp.key {
		tmp.right = bst._add(key, value, tmp.right)
	} else {
		tmp.left = bst._add(key, value, tmp.left)
	}
	return tmp
}

func (bst BST) show(tmp *Node) {
	if tmp != nil {
		bst.show(tmp.left)
		fmt.Print(tmp.key, " ")
		tmp.value.show()
		bst.show(tmp.right)
	}
}

func main() {
	bst := &BST{}
	tmp := &List{}
	tmp.add("FERNANDEZ")
	tmp.add("SANTIAGO")
	bst.add(25, tmp)
	tmp = &List{}
	tmp.add("LEMUS")
	tmp.add("KARLA")
	bst.add(5, tmp)
	tmp = &List{}
	tmp.add("CABRERA")
	tmp.add("ANGEL")
	bst.add(10, tmp)
	tmp = &List{}
	tmp.add("GONZALEZ")
	tmp.add("CARLOS")
	bst.add(41, tmp)
	bst.show(bst.root)
}
